File: libs/core/providers/implementations/bank-mock.js
import { randomUUID, createHash } from 'node:crypto';

function maybe(numberLike, fallback) {
  const n = Number(numberLike);
  return Number.isFinite(n) ? n : fallback;
}

export class MockBankEgress {
  constructor() {
    this.failureRate = Math.min(Math.max(maybe(process.env.MOCK_BANK_FAILURE_RATE, 0), 0), 1);
    this.latencyMs = Math.max(0, maybe(process.env.MOCK_BANK_LATENCY_MS, 25));
  }

  async payout(rpt, amount_cents, ref) {
    if (this.failureRate > 0 && Math.random() < this.failureRate) {
      throw new Error('Mock bank failure triggered by failure rate');
    }

    if (!rpt || !rpt.rpt_id) {
      throw new Error('RPT payload missing rpt_id');
    }

    const transferUuid = randomUUID();
    const providerReceiptId = randomUUID();
    const reference = `${ref?.periodId ?? 'unknown'}:${transferUuid}`;
    const hash = createHash('sha256').update(providerReceiptId + reference).digest('hex');

    if (this.latencyMs) {
      await new Promise((resolve) => setTimeout(resolve, this.latencyMs));
    }

    return {
      transferUuid,
      bankReceiptHash: hash,
      providerReceiptId,
      rawResponse: {
        chaos: this.failureRate,
        reference,
        amount_cents,
      },
    };
  }
}

File: libs/core/providers/implementations/bank_mock.py
"""Mock bank egress provider for Python services."""
from __future__ import annotations

import asyncio
import hashlib
import os
import uuid
from typing import Any

from libs.core.ports import PayoutReference, PayoutResult, RPT


class MockBankEgress:
    def __init__(self) -> None:
        self._latency_ms = max(0, int(os.getenv("MOCK_BANK_LATENCY_MS", "25")))
        self._failure_rate = float(os.getenv("MOCK_BANK_FAILURE_RATE", "0"))

    async def payout(self, rpt: RPT, amount_cents: int, ref: PayoutReference) -> PayoutResult:
        if self._failure_rate and self._failure_rate > 0:
            import random

            if random.random() < min(1.0, max(0.0, self._failure_rate)):
                raise RuntimeError("Mock bank failure triggered by failure rate")

        transfer_uuid = str(uuid.uuid4())
        provider_receipt_id = str(uuid.uuid4())
        reference = f"{ref.get('periodId', 'unknown')}:{transfer_uuid}"
        digest = hashlib.sha256((provider_receipt_id + reference).encode()).hexdigest()

        if self._latency_ms:
            await asyncio.sleep(self._latency_ms / 1000)

        return PayoutResult(
            transferUuid=transfer_uuid,
            bankReceiptHash=digest,
            providerReceiptId=provider_receipt_id,
            rawResponse={
                "reference": reference,
                "amount_cents": amount_cents,
                "port": "mock",
            },
        )
